### PR TITLE
FP16_Optimizer wrapper

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: FP16_Optimizer wrapper.

It is found, some version of APEX's FP16_Optimizer and DeepSpeed Optimizers 
1). are using serial overflow check, which introduces many GPU->CPU copy. 
2). are using serial gradient norm calculation, also having many GPU->CPU copy. 

Both can be addressed by simple calls, either into Torch, or Apex. This could significant decrease the latency of check overflow and clip gradient norm.  

24 layer of GPT2 Megatron-LM model, 

Without this change:

 iteration       98/     100 | elapsed time per iteration (ms): 186.1 | learning rate: 3.891E-06 | lm loss: 9.072313E+00 | loss scale: 262144.0 | number of skipped iterations:   0 | number of nan iterations:   0 |
time (ms) | forward: 57.49 | backward: 119.91 | backward-backward: 59.34 | backward-allreduce: 11.33 | backward-master-grad: 30.21 | backward-clip-grad: 18.30 | optimizer: 6.96 | batch generator: 1.37
 iteration       99/     100 | elapsed time per iteration (ms): 192.1 | learning rate: 3.937E-06 | lm loss: 9.137538E+00 | loss scale: 262144.0 | number of skipped iterations:   0 | number of nan iterations:   0 |
time (ms) | forward: 59.64 | backward: 123.78 | backward-backward: 59.00 | backward-allreduce: 13.23 | backward-master-grad: 32.41 | backward-clip-grad: 18.44 | optimizer: 6.94 | batch generator: 1.38

With this change:

iteration       98/     100 | elapsed time per iteration (ms): 144.2 | learning rate: 3.891E-06 | lm loss: 9.183796E+00 | loss scale: 262144.0 | number of skipped iterations:   0 | number of nan iterations:   0 |
time (ms) | forward: 56.70 | backward: 79.34 | backward-backward: 58.12 | backward-allreduce: 11.99 | backward-master-grad: 5.98 | backward-clip-grad: 2.72 | optimizer: 6.88 | batch generator: 1.05
iteration       99/     100 | elapsed time per iteration (ms): 149.7 | learning rate: 3.937E-06 | lm loss: 9.250105E+00 | loss scale: 262144.0 | number of skipped iterations:   0 | number of nan iterations:   0 |
time (ms) | forward: 56.72 | backward: 84.77 | backward-backward: 58.31 | backward-allreduce: 14.76 | backward-master-grad: 5.99 | backward-clip-grad: 5.09 | optimizer: 6.95 | batch generator: 1.16

elapsed time per iteration dropped ~22.38% 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
This is great performance improvement! 
Do we need some UT to guard the numerical equivalence with it's original APEX's FP16_Optimizer.

**Review by thiagocrepaldi — COMMENTED** [review]
Is there any design docavailable?

Is this going to be added to our public API?
If not, consider moving to `experimental` namespace

**Comment by pengwa** [comment]
> This is great performance improvement!
> Do we need some UT to guard the numerical equivalence with it's original APEX's FP16_Optimizer.

Good idea! We should have some tests. Will add some small ones.

**Comment by pengwa** [comment]
> Is there any design docavailable?
> 
> Is this going to be added to our public API?
> If not, consider moving to `experimental` namespace

I think we need validate this perf improvement is generic before upstreaming the changes to DeepSpeed directly. So yeah, I guess document is not needed at this phase. While you are right, we should make it a experimental namespace letting people use it with care.

File: orttraining/orttraining/python/training/ortmodule/optimizer/__init__.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# __init__.py
+
+# To make this method overriding work, please import this module at ealiest convinience for the training scripts.
+# A suggested place is the entry trainer script.
+
+from .fp16_optimizer import clip_grad_norm_fp32, check_overflow
+import nvtx
```
Comment (ytaous) [inline orttraining/orttraining/python/training/ortmodule/optimizer/__init__.py:9]: >nvtx [](http://example.com/codeflow?start=7&length=4)

remove?

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/optimizer/__init__.py:9]: Will update the code once validate deepspeed zero1 optimizer

**Comment by baijumeswani** [comment]
@pengwa would it be possible to update the usage in this pull request description (so users can reference how to use your modifier)?
